lnwallet/fee_estimator.go

```diff
@@ -474,3 +477,50 @@ type WebAPIFeeSource interface {
 	// specifics are left to the WebAPIFeeSource implementation.
 	ParseResponse(r io.Reader) (map[uint32]uint32, error)
 }
+
+// SparseConfFeeSource is an implementation of the WebAPIFeeSource that utilizes
+// a user-specified fee estimation API for Bitcoin. It expects the response
+// to be in the JSON format: `fee_by_block_target: { ... }` where the value maps
+// block targets to fee estimates (in sat per kilovbyte).
+type SparseConfFeeSource struct {
+	// URL is the fee estimation API specified by the user.
+	URL string
+}
+
+// GenQueryURL generates the full query URL. The value returned by this
+// method should be able to be used directly as a path for an HTTP GET
+// request.
+//
+// NOTE: Part of the WebAPIFeeSource interface.
+func (s SparseConfFeeSource) GenQueryURL() string {
+	return s.URL
+}
+
+// ParseResponse attempts to parse the body of the response generated by the
+// above query URL. Typically this will be JSON, but the specifics are left to
+// the WebAPIFeeSource implementation.
+//
+// NOTE: Part of the WebAPIFeeSource interface.
+func (s SparseConfFeeSource) ParseResponse(r io.Reader) (map[uint32]uint32, error) {
+	kresp, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+
+	type jsonResp struct {
+		FeeByBlockTarget map[uint32]uint32 `json:"fee_by_block_target"`
+	}
+
+	resp := jsonResp{FeeByBlockTarget: make(map[uint32]uint32)}
+	dec := bytes.NewReader(kresp)
```

Comment:
why not use `r` directly?

Reply:
wow, this improved the function a lot. fixed